### PR TITLE
Client side data processing

### PR DESCRIPTION
**Pull request by serhiipylypchuk1991**

### CODE REVIEW

File: js/index.js [hunk 1]
```diff
@@ -58,18 +62,22 @@ webix.ready(function(){
     minWidth:400,
     scheme:{
         $init:function(obj){
+            obj.categoryId = getRandomInt(1,4);
+            obj.rank = this.count();//Like variant
```
Comment (Zwillinge) [inline js/index.js]: looks strange. $init applied to loaded data as well, so you just corrupt the initial data. you can check `if(!obj.rank)` for example. but why this line is necessary at all?

File: js/index.js [hunk 2]
```diff
@@ -215,107 +216,161 @@ webix.ready(function(){
...
+    height:300,
+    scrollY:true,
+    scrollX:false,
+    select:true,
+    rules:{
+      name:webix.rules.isNotEmpty
+    },
+    onClick:{
+      remove_list_item_btn:function(e, id){
+        this.remove(id);
```
Comment (Zwillinge) [inline js/index.js]: it would be nice to have a confirm message for remove operations

File: js/index.js [hunk 3]
```diff
@@ -215,107 +216,161 @@ webix.ready(function(){
...
+      end:10,
+      step:2
+    }
   };
   var treetable = {
     view:"treetable",
+    //id:"tree",
+    scheme:{
+      $init:function(obj){
+        obj.open = true;
```
Comment (Zwillinge) [inline js/index.js]: not very good, as you assign this property to every item, not just branches. you can open them after loading as

```
ready:function(){
    this.openAll();
}
```

File: js/index.js [hunk 4]
```diff
@@ -215,107 +216,161 @@ webix.ready(function(){
...
     scrollX:false,
     select:true,
     columns:[
       { id:"id", header:"", css:"treetable_id", width:50},
-      { id:"title",	header:"Title", css:"treetable_title", minWidth:300, template:"{common.treetable()} #title#"},
-      { id:"price",	header:"Price", minWidth:300, css:"treetable_price", fillspace:true}
+      { id:"title",	header:"Title", editor:"text", css:"treetable_title", minWidth:300, liveEdit:true, template:"{common.treetable()} #title#"},
+      { id:"price",	header:"Price", editor:"text", minWidth:300, css:"treetable_price", liveEdit:true, fillspace:true}
     ],
+    on:{
```
Comment (Zwillinge) [inline js/index.js:336]: first-level items are also edited and validation fails (expected)
to cancel editing to particular items, use the onBeforeEditStart event. inside it you can get the data item and check its `$count` property. this helper property stores the number of kids, and it is 0 for branches

**Review by Zwillinge — CHANGES_REQUESTED** [review]

File: js/index.js (hunk 1 above)
Comment (serhiipylypchuk1991) [reply js/index.js]: In situation, when you add new element, "rank" cell will be empty without this line. (obj.rank = this.count())

File: js/index.js (hunk 2 above)
Comment (serhiipylypchuk1991) [reply js/index.js]: Confirmation was added

File: js/index.js (hunk 3 above)
Comment (serhiipylypchuk1991) [reply js/index.js]: It was changed

File: js/index.js (hunk 4 above)
Comment (serhiipylypchuk1991) [reply js/index.js:336]: Using this event, I can just cancel editing data of 1 level (whole row)! How to cancel editing data just in 1 cell (empty cell) of this row?

Comment (Zwillinge) [reply js/index.js:336]: id is in object that stores cell coordinates `{row: 2, column: "title"}`, so you can check for column id as well

Comment (serhiipylypchuk1991) [reply js/index.js:336]: Problem is solved by next code:
onBeforeEditStart:function(id){
        if ( this.getItem(id).$level === 1 && id.column === "price"){
          return false;
        }
      },
      onValidationError:function(id, obj, details){
        if(details.price === true){
          return false;
        }
      }

**Review by Zwillinge — APPROVED** [review]

**Pull request merged**